elastica/rigidbody/mesh_rigid_body.py

```diff
@@ -0,0 +1,193 @@
+                    self.faces[:, i, k] - center_of_mass
+                ) / self.distance_to_faces_from_center_of_mass[
+                    i, k
+                ]
+
+        self.face_normals_in_material_frame = np.zeros(
+            (MaxDimension.value(), self.n_faces)
+        )
+        self.face_normals_in_material_frame = self.face_normals.copy()
+
+        # position is at the center of mass
+        self.position_collection = np.zeros((MaxDimension.value(), 1))
+        self.position_collection[:, 0] = center_of_mass
+
+        self.velocity_collection = np.zeros((MaxDimension.value(), 1))
+        self.omega_collection = np.zeros((MaxDimension.value(), 1))
+        self.acceleration_collection = np.zeros((MaxDimension.value(), 1))
+        self.alpha_collection = np.zeros((MaxDimension.value(), 1))
+
+        self.external_forces = np.zeros((MaxDimension.value())).reshape(
+            MaxDimension.value(), 1
+        )
+        self.external_torques = np.zeros((MaxDimension.value())).reshape(
+            MaxDimension.value(), 1
+        )
+
+    def update_faces(self):
+        _update_faces(
+            self.director_collection,
+            self.faces,
+            self.position_collection,
+            self.distance_to_faces_from_center_of_mass,
+            self.direction_to_faces_from_center_of_mass_in_material_frame,
+            self.n_faces,
+        )
+        _update_face_centers(
+            self.director_collection,
+            self.face_centers,
+            self.position_collection,
+            self.distance_to_face_centers_from_center_of_mass,
+            self.direction_to_face_centers_from_center_of_mass_in_material_frame,
+            self.n_faces,
+        )
+        _update_face_normals(
+            self.director_collection,
+            self.face_normals,
+            self.face_normals_in_material_frame,
+            self.n_faces,
+        )
+
+
+@numba.njit(cache=True)
+def _update_face_centers(
+    director_collection,
+    face_centers,
+    center_of_mass,
+    distance_to_face_centers_from_center_of_mass,
+    direction_to_face_centers_from_center_of_mass_in_material_frame,
+    n_faces,
+):
+    face_centers[:] = np.zeros((3, n_faces))
+    for k in range(n_faces):
+        for i in range(3):
+            face_centers[i, k] += center_of_mass[i, 0]
+            for j in range(3):
+                face_centers[i, k] += (
+                    distance_to_face_centers_from_center_of_mass[i]
+                    * director_collection[i, j, 0]
+                    * direction_to_face_centers_from_center_of_mass_in_material_frame[
+                        j, k
+                    ]
+                )
+
+
+@numba.njit(cache=True)
+def _update_faces(
+    director_collection,
+    faces,
+    center_of_mass,
+    distance_to_faces_from_center_of_mass,
+    direction_to_faces_from_center_of_mass_in_material_frame,
+    n_faces,
+):
+    faces[:] = np.zeros((3, 3, n_faces))  # dim,vertices,faces
```

Comment:
What is the difference between these two functions? It seems there is common code and computation, can we merge them? And can you provide a little bit documentation.

Reply:
Merged and added comments